### PR TITLE
calli with default cc should not be prevented from being considered as an inline candidate

### LINKED ISSUE #12713: Methods with calli fail to inline because of unsupported opcode
Today an occurrence of the calli opcode in a method body prevents it from being even considered as an inline candidates quite early on in the Flowgraph processing, while finding basic blocks and then jump targets.

Specifically in Compiler::fgFindJumpTargets(https://github.com/dotnet/coreclr/blob/master/src/jit/flowgraph.cpp#L4767)

>
            case CEE_CALLI:
            case CEE_LOCALLOC:
            case CEE_MKREFANY:
            case CEE_RETHROW:
                // CEE_CALLI should not be inlined because the JIT cannot generate an inlined call frame. If the call
                // target
                // is a no-marshal CALLI P/Invoke we end up calling the IL stub. We don't NGEN these stubs, so we'll
                // have to
                // JIT an IL stub for a trivial func. It's almost certainly a better choice to leave out the inline
                // candidate so we can generate an inlined call frame. It might be nice to call getCallInfo to figure
                // out
                // what kind of call we have here.

This limitation (about inlined call frame for PInvokes) may still hold true for unmanaged calling conventions that require it, but for default calling convention targets there is no need to raise a frame (afaict).

We should be able to remove the restriction for call targets with the default cc without any major plumbing.

### PR DESCRIPTION
**Pull request by mjsabby**
Fixes #12713.

### CODE REVIEW

File: src/jit/flowgraph.cpp [hunk 1]
```diff
@@ -4761,19 +4761,37 @@ void Compiler::fgFindJumpTargets(const BYTE* codeAddr, IL_OFFSET codeSize, BYTE*
...
+                // CORINFO_CALLCONV_DEFAULT. In the case where we have a no-marshal CALLI P/Invoke we end up calling
+                // the IL stub. We don't NGEN these stubs, so we'll have to JIT an IL stub for a trivial func.
+                // It's almost certainly a better choice to leave out the inline candidate so we can generate an inlined 
+                // call frame.
+
+                // Consider making this only for not force inline.
+                if (makeInlineObservations)
+                {
+                    CORINFO_SIG_INFO sig;
+                    info.compCompHnd->findSig(info.compScopeHnd, getU4LittleEndian(codeAddr),
```
Comment (AndyAyersMS) [inline src/jit/flowgraph.cpp]: Probably should add the paranoia check here that there are actually enough bytes of IL left to read the token.

File: src/jit/flowgraph.cpp [hunk 2]
```diff
@@ -4761,19 +4761,37 @@ void Compiler::fgFindJumpTargets(const BYTE* codeAddr, IL_OFFSET codeSize, BYTE*
...
+                                              impTokenLookupContextHandle, &sig);
+
+                    if (sig.getCallConv() != CORINFO_CALLCONV_DEFAULT)
```
Comment (AndyAyersMS) [inline src/jit/flowgraph.cpp]: It likely doesn't matter, but `findSig` does a lot more work than just finding the calling convention. Would be good to verify that this is not a particularly hot path and the extra cost here doesn't show up in profiles.

**Review by AndyAyersMS — COMMENTED** [review]
Are there existing test cases that cover this change? If not, can you add some (both positive and negative)? 

It would be good to verify that if we allow these inlinees, there aren't other blockers elsewhere.

Also we probably should run this through desktop testing before merging.

File: src/jit/flowgraph.cpp (hunk 2 above)
Comment (mjsabby) [inline src/jit/flowgraph.cpp]: Is there an alternative to findSig you know off hand? I'll look around to see.

**Comment by mjsabby** [comment]
@AndyAyersMS My PR is certainly having an effect on some tests. They're all failing on assert(!compIsForInlining) which I'm assuming indicates if the compiler was invoked for inlining or not, but I'm not sure why it'll be illegal now with my change.

**Comment by AndyAyersMS** [comment]
The assert comes from the fact that the case for `CEE_JMP` now falls through to `CEE_CALLI`. You should reorder these two cases so that `CEE_JUMP` falls through to right logic.

**Comment by AndyAyersMS** [comment]
Changes look good. Can you run jit-diffs on this?

**Comment by mjsabby** [comment]
@AndyAyersMS yup, I'll figure out how to do that today.

**Comment by karelz** [comment]
What is status of this PR? No update in last month :( ...

**Comment by AndyAyersMS** [comment]
I need to run some desktop testing on this for @mjsabby ... will try to get to it next week.

**Comment by AndyAyersMS** [comment]
@mjsabby what is the intent of the dynamic module check?

**Comment by AndyAyersMS** [comment]
I went ahead and removed the dynamic module check.

Desktop testing mostly good.  Have one test failure I need to look at more closely, but suspect it is something off with my destkop test environment. So am scorching it all and will rerun the tests.

No diffs seen in desktop SPMI. Likely this pattern is only reachable in IL based tests or equivalent. I'll add a test case to CoreCLR where an inline that was previously blocked is now allowed.

**Comment by AndyAyersMS** [comment]
I have some updates to the code so am going to close this and resubmit from my fork.

**Comment by mjsabby** [comment]
Sorry for not responding here ealirer, had to get some family stuff sorted out. the dynamic check was there because it seemed like calling findSig on the dynamic method failed with an ASSERT. But I suppose things are good if it passed desktop testing.

**Pull request closed without merging**